pallets/identity/src/keys.rs

```diff
@@ -244,6 +261,75 @@ impl<T: Config> Module<T> {
         Ok(())
     }
 
+    /// Accepts a primary key rotation.
+    /// Differs from accept_primary_key_rotation in that it will leave the old primary key as a
+    /// secondary key with the permissions specified in the corresponding DirectRotatePrimaryKey authorization
+    /// instead of unlinking the primary key.
+    crate fn base_rotate_primary_key_to_secondary(
+        origin: T::Origin,
+        rotation_auth_id: u64,
+        optional_cdd_auth_id: Option<u64>,
+    ) -> DispatchResult {
+        let new_primary_key = ensure_signed(origin)?;
+        let new_primary_key_signer = Signatory::Account(new_primary_key.clone());
+        Self::accept_auth_with(
+            &new_primary_key_signer,
+            rotation_auth_id,
+            |data, target_did| {
+                Self::ensure_key_unlinked_or_secondary_key(&new_primary_key, &target_did)?;
+
+                let perms = extract_auth!(data, RotatePrimaryKeyToSecondary(p));
+
+                // Accept authorization from CDD service provider.
+                if Self::cdd_auth_for_primary_key_rotation() {
+                    let auth_id = optional_cdd_auth_id
+                        .ok_or_else(|| Error::<T>::InvalidAuthorizationFromCddProvider)?;
+
+                    let signer = Signatory::Account(new_primary_key.clone());
+                    Self::accept_auth_with(&signer, auth_id, |data, auth_by| {
+                        let attestation_for_did = extract_auth!(data, AttestPrimaryKeyRotation(a));
+                        // Attestor must be a CDD service provider.
+                        ensure!(
+                            T::CddServiceProviders::is_member(&auth_by),
+                            Error::<T>::NotCddProviderAttestation
+                        );
+                        // Ensure authorizations are for the same DID.
+                        ensure!(
+                            target_did == attestation_for_did,
+                            Error::<T>::AuthorizationsNotForSameDids
+                        );
+                        Ok(())
+                    })?;
+                }
+
+                let mut record = DidRecords::<T>::get(target_did);
+
+                let old_primary_key = record.primary_key.clone();
+
+                // Remove old secondary key
+                let removed_signers = vec![new_primary_key_signer.clone()];
+                record.remove_secondary_keys(&removed_signers);
+                Self::deposit_event(RawEvent::SecondaryKeysRemoved(target_did, removed_signers));
```

Comment:
Removing a new_primary_key from secondary keys is only required if the key was one.  
Suggest to adding `if Self::is_signer(target_did, &new_primary_key_signer) {` check to these steps.
or maybe since the did records have already been retrieved above just check directly if it is a signer
`if record.secondary_keys.iter().any(|si| si.signer == &new_primary_key_signer) {`

Reply:
Removal of secondary key, if if is one, now also needs to be added to `unsafe_primary_key_rotation` so it might be best to make it a separate function.